gpustack/policies/candidate_selectors/gguf_resource_fit_selector.py

```diff
@@ -1622,144 +1583,183 @@ async def _generate_combinations_with_selected_gpus(
 
-            # Skip the worker if the main worker can't offload even one layer.
-            main_worker = worker_map.get(r[0])
-            if not self._can_offload_at_least_one_layer(
-                r[1],
-                self._get_single_layer_vram(
-                    main_worker.status.memory.is_unified_memory
-                ),
-            ):
-                return None, None, None
+        worker_names = list(self._selected_gpu_ids_by_worker.keys())
+        worker_ids = [worker.id for worker in workers if worker.name in worker_names]
+        for w in workers_allocatable_vram:
+            if w[0] in worker_ids:
+                selected_workers_allocatable_vram.append(w)
 
-            c = [(r, *[gpu for gpu in sorted_gpus if gpu[0] != r[0]])]
+        selected_workers_gpus_allocatable_vram = []
+        for selected_gpu_id in self._selected_gpu_ids:
+            valid, matched = parse_gpu_id(selected_gpu_id)
+            if not valid:
+                continue
 
-            for item in c:
-                key = len(item)
-                if key not in combinations:
-                    combinations[key] = []
+            selected_worker_name = matched.get("worker_name")
+            selected_gpu_index = matched.get("gpu_index")
+            selected_worker_id = worker_name_id_map.get(selected_worker_name)
+            for w in workers_gpus_allocatable_vram:
+                if w[0] == selected_worker_id and str(w[1]) == selected_gpu_index:
+                    selected_workers_gpus_allocatable_vram.append(w)
 
-                combinations[key].append(item)
+        sorted_workers, sorted_gpus = _sort_and_group_worker_gpu_vram(
+            selected_workers_allocatable_vram,
+            selected_workers_gpus_allocatable_vram,
+        )
 
-            logger.debug(
-                f"Generated combinations with main: {main_worker.name} and rpcs number: {len(c) - 1}"
+        main_worker_vram, main_worker = self._get_main_for_combination(
+            sorted_workers, workers_allocatable, worker_map
+        )
+
+        if not main_worker_vram:
+            return None
+
+        # Check if the rpc gpus can offload even one layer.
+        filtered_gpus = [gpu for gpu in sorted_gpus if gpu[0] != main_worker_vram[0]]
+
+        for gpu in filtered_gpus:
+            rpc_at_least_vram = self._get_single_layer_vram(
+                worker_map.get(gpu[0]).status.memory.is_unified_memory, True
             )
-        else:
-            if (
-                len(workers_allocatable_vram) == 0
-                or len(workers_gpus_allocatable_vram) == 0
+
+            if not self._can_offload_at_least_one_layer(
+                gpu[2],
+                rpc_at_least_vram,
             ):
-                return None, None, None
+                key = f"{worker_map.get(gpu[0]).name}:{gpu[2]}"
+                logger.warning(f"Selected gpu {key} can't offload at least one layer")
```

Comment:
Can users know why scheduling fails if they don't view the logs?

Reply:
More information can be provided in this issue [849](https://github.com/gpustack/gpustack/issues/849).